packages/aws-client/src/utils.ts

```diff
@@ -68,3 +68,34 @@ export const retryOnThrottlingException = <T, U extends unknown[]>(
         () => fn(...args).catch(retryIfThrottlingException),
         { maxTimeout: 5000, ...options }
       );
+
+const retryIfMissingObjectError = (error: Error & { statusCode?: number }) => {
+  const { statusCode } = error;
+
+  if (statusCode && [404, 412].includes(statusCode)) throw error;
+  throw new pRetry.AbortError(error);
+};
+
+/**
+ * Returns a function that wraps the specified function, but will retry the
+ * wrapped function when it throws a "not found" (404) or "pre-condition failed"
+ * (412) error, based upon the specified retry options.
+ *
+ * @param {Function} fn - function (that returns a Promise) to retry on error
+ * @param {Object} options - retry options
+ * @returns {Function} a function that will retry the specified function when
+ *    it throws a "not found" or "pre-condition failed" error (i.e., the
+ *    `statusCode` property of the error is either 404 or 412)
+ * @see https://github.com/sindresorhus/p-retry#options
+ * @see https://github.com/tim-kos/node-retry#retryoperationoptions
+ * @see https://github.com/tim-kos/node-retry#retrytimeoutsoptions
+ */
+export const retryOnMissingObjectError = <T, U extends unknown[]>(
```

Comment:
Is this actually used? Do we need it? I'd rather not add it if we're not using it, especially given that the signature is more than a little complex to understand

Reply:
I see that it is used later. Given the complexity of this function I would vastly prefer putting the `p-retry` directly inside `getObject`

Reply:
Yes, I added it because I use it on other code in this PR. Regarding the signature, it is identical to the `retryOnThrottlingException` that is already part of this file.

However, this does tie back into my other comment about `pRetry` being a bit clumsy and this was my compromise in favor of the more general convenience wrapper that I decided against adding to avoid a long discussion on this PR and save it for an Arch. Mtg.

Reply:
Can you elaborate on what you find complex about it? Keep in mind, I copied the existing `retryOnThrottlingException` function in this file, so there is precedence for it.

For comparison, a bare call to `getObject` would look like so:
```
const obj = getObject(s3(), { Bucket, Key });
```
Adding retries using the function in question might look like so (again, similar to usage of `retryOnThrottlingException`):
```
const retryGetObject = retryOnMissingObjectError(getObject, { retries: 5 });
const obj = retryGetObject(s3(), { Bucket, Key });
```

More generally, regardless of whether you like this particular function or not (I'm actually not a fan of it myself -- it's a compromise I made -- which ties into my desire to discuss this topic in an Arch. Mtg.), decoupling the retry capabilities from `getObject` (as well as the other S3 functions) is important because we may not always want to retry on the same set of exceptions.

If we bake the retry capabilities into the `getObject` function, we cannot change which exceptions to retry on, unless we add _further_ complexity to the function signature, beyond the addition of the retry options object.

For example, in one case, we might want to retry _only_ on pre-condition failures, and _not_ on "no such bucket/key" failures, or vice versa, or even for something else, like "not modified."